buffer/src/main/java/io/netty/buffer/BufferOutputStream.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer;
+
+import io.netty.buffer.api.Buffer;
+import io.netty.util.CharsetUtil;
+
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * An {@link OutputStream} which writes data to a {@link Buffer}.
+ * <p>
+ * A write operation against this stream will occur at the {@code writerOffset}
+ * of its underlying buffer and the {@code writerOffset} will increase during
+ * the write operation.
+ * <p>
+ * This stream implements {@link DataOutput} for your convenience.
+ * The endianness of the stream is always big endian.
+ *
+ * @see BufferInputStream
+ */
+public class BufferOutputStream extends OutputStream implements DataOutput {
+    private final Buffer buffer;
+    private final int startIndex;
+    private DataOutputStream utf8out; // lazily-instantiated
+    private boolean closed;
+
+    /**
+     * Creates a new stream which writes data to the specified {@code buffer}.
+     */
+    public BufferOutputStream(Buffer buffer) {
+        this.buffer = requireNonNull(buffer, "buffer");
+        startIndex = buffer.writerOffset();
+    }
+
+    /**
+     * Returns the number of written bytes by this stream so far.
+     */
+    public int writtenBytes() {
+        return buffer.writerOffset() - startIndex;
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        if (len == 0) {
```

Comment:
we should also throw if len is 0 but the stream was closed

Reply:
The super-class, `OutputStream`, does not do that.

Reply:
Whole this is true I guess we should still do this as its also done by other implementations in the JDK: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/io/OutputStream.java#L89